---
title: "Filter by predicted variant effect on protein"
output: html_document
editor_options: 
  chunk_output_type: console
---

started: AL25Sep2018  
last updated: AL19Sep2019  

# Summary  

Input: WECARE-NFE dataset of 1,309 variants x 664 samples (467BC = 232UBC + 235CBC and 197NFE)
Output: 162 variants x 664 samples  

First, select variants that are either:  
- LoF (frameshift, stop/start gain/loss, splise donor/acceptor)  
- Missence that are simulteneously **deleterious** by SIFT *AND* **probably_damaging** by PolyPhen  
- Known pathogenic/ likely-pathogenic or risk factors by ClnSig (from the old VEP, compartible with WES)  

Then exclude variants that are known **benign** or **likely benign** by ClinSig  

Explore remaining variants  

Explore presence of genes of interest:  
"ATM","CHEK2","NF1","NRG1","ERCC6", "FOXM1" and "SLC9A3R1"  

# start_section

<style>
pre {
  overflow-x: auto;
}
pre code {
  word-wrap: normal;
  white-space: pre;
}
</style>

```{r set_width, echo=F}

#eval=F

options(width = 999)
# https://stackoverflow.com/questions/36845178/width-of-r-code-chunk-output-in-rmarkdown-files-knitr-ed-to-html

```

```{r start_section}

Sys.time()
rm(list=ls())
graphics.off()

library(knitr)
library(dplyr)
library(ggplot2)
library(VennDiagram)

base_folder="/Users/alexey/Documents/wecare/ampliseq/v04_ampliseq_nfe/s15_filter_by_effect_on_protein_and_AF"

opts_knit$set(root.dir = base_folder)

options(stringsAsFactors = F)
options(warnPartialMatchArgs = T, 
        warnPartialMatchAttr = T, 
        warnPartialMatchDollar = T)

#options(error = browser()) # Type Q or c to exit, drop browser level
# https://support.rstudio.com/hc/en-us/articles/200713843?version=1.1.456&mode=desktop
# https://stackoverflow.com/questions/13052522/how-to-leave-the-r-browser-mode-in-the-console-window/13052588 

```

# read_data  

```{r read_data}

genes_107 <- scan(
  file="/Users/alexey/Documents/wecare/ampliseq/a01_genes_from_EF/candidate_genes_ampliseq_107G_EF_20171122.txt",
  what="character")

load(paste(base_folder, "s01_filter_by_AF.RData", sep="/"))

```

# Check data

```{r start_check}

dim(genotypes.mx)
dim(phenotypes.df)
dim(variants.df)

sum(rownames(variants.df) != rownames(genotypes.mx))
sum(colnames(genotypes.mx) != rownames(phenotypes.df))

length(genes_107)

```

# --- Explore_variants_annotations --- #

## Check genes

Low-frequency variants were detected in 90 genes, 87 of them were expected and 3 - unexpected:  

- It seems that a typo in gene name happened during the panel design: *DCLRE1C* was included instead of CDLRE1C  
- One gene was detected under a different name: REV7 was detected as *MAD2L2*  
- Some variants were mapped to one unintended gene: *PI4KA*. This gene overlaps with SERPIND1, included in the panel.  

No variants were detected in 20 targeted genes (including REV7=MAD2L2, XPF=ERCC4)  


```{r check_genes, width=80}

detected_genes <- unique(variants.df$SYMBOL)
length(detected_genes)
sort(detected_genes)

detected_expected <- intersect(detected_genes, genes_107)
length(detected_expected)
sort(detected_expected)

detected_unexpected <- setdiff(detected_genes, genes_107)
length(detected_unexpected)
sort(detected_unexpected)

missed <- setdiff(genes_107, detected_genes)
length(missed)
sort(missed)

rm(detected_genes, detected_expected, detected_unexpected, missed)

```

## Available variant annotations

```{r annotations_list}

colnames(variants.df)

```

## VEP_consequences

A combination of consequencies can be assigned to a variant  

```{r VEP_consequences}

sum(is.na(variants.df$Consequence)) # 0
a <- strsplit(as.vector(variants.df$Consequence),"&")
sort(table(unlist(a)), decreasing=T)
rm(a)

```

## sift, polyphen

```{r explore_SIFT_PolyPhen}

# SIFT
sum(is.na(variants.df$SIFT_call))
sort(table(as.vector(variants.df$SIFT_call)),decreasing = T)

# PolyPhen
sum(is.na(variants.df$PolyPhen_call))
sort(table(as.vector(variants.df$PolyPhen_call)),decreasing = T)

```

## VEP Clin_Sig

A combination of consequencies can be assigned to a variant  

```{r clin_sig}

sum(is.na(variants.df$CLIN_SIG))
sort(table(unlist(strsplit(as.vector(variants.df$CLIN_SIG), split="\\&"))),decreasing = T)

```

## Multiallelic varaints

<3% of all variants  

```{r multiallelic_varaints}

summary(as.factor(variants.df$Multiallelic))
36/(1273+36)

```

# --- Make indices --- #

## make_lof_index

frameshift, splice donor/acceptor, stop gain, start lost

```{r make_lof_index}

# Make sure there is no NAs in consequence
sum(is.na(variants.df$Consequence)) # 0

# --- Frameshifts --- #

frameshift_variant <- grepl( "frameshift_variant", variants.df$Consequence)
sum(frameshift_variant)

# mean count of alleles per case in wecare
fsv.mx <- genotypes.mx[frameshift_variant,]
fsv.counts <- apply(fsv.mx, 2, sum, na.rm=TRUE)
fsv.mean <- mean(fsv.counts, na.rm=TRUE) 
fsv.mean
rm(fsv.mx, fsv.counts)

# --- Splice acceptor variants (sav) --- #

splice_acceptor_variant <- grepl( "splice_acceptor_variant", variants.df$Consequence)
sum(splice_acceptor_variant)

# mean count of alleles per case
sav.mx <- genotypes.mx[splice_acceptor_variant,]
sav.counts <- apply(sav.mx, 2, sum, na.rm=TRUE)
sav.mean <- mean(sav.counts, na.rm=TRUE)
sav.mean
rm(sav.mx, sav.counts)

# --- Splice donor variants (sdv) --- #

splice_donor_variant <- grepl( "splice_donor_variant", variants.df$Consequence)
sum(splice_donor_variant)

# mean count of alleles per case in wecare
sdv.mx <- genotypes.mx[splice_donor_variant,]
sdv.counts <- apply(sdv.mx, 2, sum, na.rm=TRUE)
sdv.mean <- mean(sdv.counts, na.rm=TRUE) 
sdv.mean
rm(sdv.mx, sdv.counts)

# --- Stop gains --- #

stop_gain <- grepl( "stop_gain", variants.df$Consequence)
sum(stop_gain)

# mean count of alleles per case in wecare
stogv.mx <- genotypes.mx[stop_gain,]
stogv.counts <- apply(stogv.mx, 2, sum, na.rm=TRUE)
stogv.mean <- mean(stogv.counts, na.rm=TRUE) 
stogv.mean
rm(stogv.mx, stogv.counts)

# --- Start losts --- #

start_lost <- grepl( "start_lost", variants.df$Consequence)
sum(start_lost)

# mean count of alleles per case in wecare
stalv.mx <- genotypes.mx[start_lost,]
stalv.counts <- apply(stalv.mx, 2, sum, na.rm=TRUE)
stalv.mean <- mean(stalv.counts, na.rm=TRUE) 
stalv.mean
rm(stalv.mx, stalv.counts)

# One of the start losts is quite common (~4.6%): it is kept, but will be down-weighted later
variants.df[start_lost,c("SYMBOL","Consequence","CLIN_SIG","total_AF")]

# --- all LoF-s --- #
# Note yet excluding known benign !

lof <- frameshift_variant | splice_acceptor_variant | splice_donor_variant | stop_gain | start_lost
sum(lof)

# mean count of alleles per case in wecare
lof.mx <- genotypes.mx[lof,]
lof.counts <- apply(lof.mx, 2, sum, na.rm=TRUE)
lof.mean <- mean(lof.counts, na.rm=TRUE) 
lof.mean
rm(lof.mx, lof.counts)

# ------- Plot average counts of lof-s per individual ------- #

# Summary table with average numbers of lof-s per individual
lof_summary_counts.df <- as.data.frame(cbind(
  lof_type=c("all_LoFs","start_lost","stop_gain","frameshift","splice_acceptor","splice_donior"),
  lof_count=c(lof.mean,stalv.mean,stogv.mean,fsv.mean,sav.mean,sdv.mean)))
str(lof_summary_counts.df)
lof_summary_counts.df$lof_count <- as.numeric(as.vector(lof_summary_counts.df$lof_count))
str(lof_summary_counts.df)

# Order of levels defines order of bars on the plot
lof_summary_counts.df$lof_type <- factor(lof_summary_counts.df$lof_type, 
  levels=c("all_LoFs", "start_lost", "stop_gain", "frameshift", "splice_acceptor", "splice_donior"))

# Plot
ggplot(lof_summary_counts.df, aes(lof_type, lof_count)) +
  geom_bar(stat="identity", fill=c("red",rep("blue",5)), alpha=0.5) +
  geom_text(aes(label=round(lof_count,3)), vjust=-1, size=4) +
  labs(x = "", y = "") + 
  ylim(0,0.2) +
  theme(axis.text.x = element_text(hjust=1, size=18, angle=45)) + 
  ggtitle("Average counts of rare and low-frequency LoF variants\nper individual (ampliseq-nfe, MAF<0.05)")

# Clean-up
rm(splice_acceptor_variant, splice_donor_variant, stop_gain, frameshift_variant, 
   start_lost, sav.mean, sdv.mean, stogv.mean, stalv.mean, fsv.mean, 
   lof.mean, lof_summary_counts.df)

```

## DDM : Deleterious Damaging Missense

missense variants that are both: [ deleterious by SIFT ] AND [ probably_damaging by PolyPhen ]  

NB: Not yet excluding known benign/contradicting

```{r make_missense_index}

missense_all <- grepl( "missense_variant", variants.df$Consequence)
sum(missense_all)

# Note many NAs: these are when SIFT is irrelevant (e.g. non-coding)
# Note substitution of NAs to FALSE: this is required for combining logical vectors later 
# Importantly, the meaning of NAs is equivalent to FALSE in the given context

summary(as.factor(variants.df$SIFT_call))
deleterious <- variants.df$SIFT_call == "deleterious"
summary(deleterious)
FALSE -> deleterious[is.na(deleterious)]
summary(deleterious)

summary(as.factor(variants.df$PolyPhen_call))
probably_damaging <- variants.df$PolyPhen_call == "probably_damaging"
summary(probably_damaging)
FALSE -> probably_damaging[is.na(probably_damaging)]
summary(probably_damaging)

ddm <- deleterious & probably_damaging
sum(ddm)

# Plot venn diagram of selected variants by types
grid.newpage()
draw.pairwise.venn(
  area1=sum(deleterious), 
  area2=sum(probably_damaging), 
  cross.area=sum(ddm), 
  category=c(
    paste("SIFT\ndeleterious\n", sum(deleterious)),
    paste("PolyPhen\nprobably_damaging\n", sum(probably_damaging))),
  fill=c("red", "blue"), 
  alpha = c(0.3, 0.3),
  cex=2)

# Clean-up
rm(missense_all, deleterious, probably_damaging)

```

## VEP_ClinSig

The VEP ClinSig data are used here for comparability with WES analysis.  

*VEP *CLIN_SIG* annotations found in the file*  
pathogenic/likely_pathogenic, risk_factor, 
benign/likely_benign, uncertain_significance, not_provided  

*Pathogenic VEP-clinsig variants, which will be selected*  
pathogenic/likely_pathogenic, risk_factor  

*Benign VEP-clinsig variants, which will be excluded later*  
likely_benign/benign  

Inconsistent annotations (pathogenic + benign) will be excluded  
because excluding benigns follows after selecting pathogenics  

Note that there is no NA in the logic vectors produced by grepl()  
so, these vectors can be combined without removing NA and  
sum function calls would not need na.rm=TRUE  

###VEP ClinSig pathogenic variants  

```{r vep_clinsig}

# A combination of consequencies can be assigned to a variant  
sum(is.na(variants.df$CLIN_SIG))
sort(table(unlist(strsplit(as.vector(variants.df$CLIN_SIG), split="\\&"))), decreasing = T)

# --- likely_pathogenic and pathogenic (pat)  --- #

vep_pathogenic_variant <- grepl( "pathogenic", variants.df$CLIN_SIG)
summary(vep_pathogenic_variant)

# mean count of alleles per case 
vep_pat.mx <- genotypes.mx[vep_pathogenic_variant,]
vep_pat.counts <- apply(vep_pat.mx, 2, sum, na.rm=TRUE)
vep_pat.mean <- mean(vep_pat.counts, na.rm=TRUE)
vep_pat.mean
rm(vep_pat.mx, vep_pat.counts)

# --- risk_factor (rsf)  --- #

vep_risk_factor_variant <- grepl( "risk_factor", variants.df$CLIN_SIG)
summary(vep_risk_factor_variant)

# mean count of alleles per case: single variant
vep_rsf.mx <- genotypes.mx[vep_risk_factor_variant,,drop=F]
vep_rsf.counts <- apply(vep_rsf.mx, 2, sum, na.rm=TRUE)
vep_rsf.mean <- mean(vep_rsf.counts, na.rm=TRUE)
vep_rsf.mean
rm(vep_rsf.mx,vep_rsf.counts)

# --- Make joined index of pathogenic VEP clinsig --- #
vep_clinsig <- vep_pathogenic_variant | vep_risk_factor_variant
summary(vep_clinsig)

# mean count of alleles per case 
vep_clinsig.mx <- genotypes.mx[vep_clinsig,]
vep_clinsig.counts <- apply(vep_clinsig.mx, 2, sum, na.rm=TRUE)

table(vep_clinsig.counts) # 2 cases with 5 pathogenic clinsig variants per case !
colnames(vep_clinsig.mx)[vep_clinsig.counts==5]

vep_clinsig.mean <- mean(vep_clinsig.counts, na.rm=TRUE)
vep_clinsig.mean
rm(vep_clinsig.mx, vep_clinsig.counts)

# Clean-up
rm(vep_pathogenic_variant, vep_risk_factor_variant)

```

### Plot average counts of pathogenic VEP ClinSig variants per individual  

```{r plot_vep_pat_clinsig}

# Summary table with average numbers of clinsig variants per individual
vep_clinsig_counts.df <- as.data.frame(cbind(
  clinsig_type=c("pathogenic", "risk_factor", "All_VEP_clinsig"),
  clinsig_count=c(vep_pat.mean, vep_rsf.mean, vep_clinsig.mean)))

# Convert to vector and sort
vep_clinsig_counts.df$clinsig_count <- as.numeric(as.vector(vep_clinsig_counts.df$clinsig_count))
vep_clinsig_counts.df <- vep_clinsig_counts.df %>% arrange(desc(clinsig_count))

# Order of levels defines order of bars on the plot
vep_clinsig_counts.df$clinsig_type <- factor(vep_clinsig_counts.df$clinsig_type, 
  levels=c("All_VEP_clinsig","pathogenic","risk_factor"))

# Plot
ggplot(vep_clinsig_counts.df, aes(clinsig_type, clinsig_count)) +
  geom_bar(stat="identity", fill=c("red",rep("blue",2)), alpha=0.5) +
  geom_text(aes(label=round(clinsig_count,3)), vjust=-1, size=5) +
  labs(x = "", y = "") + 
  ylim(0, 0.07) + 
  theme(axis.text.x = element_text(hjust=1, size=18, angle=45)) +
  ggtitle("Mean counts of rare and low-frequency VEP-ClinSig pathogenic variants\nper individual (MAF<0.05, risk factor is inclusive in pathogenic)")

# Clean-up
rm(vep_pat.mean, vep_rsf.mean, vep_clinsig.mean)

```

###VEP ClinSig benign variants  

```{r vep_clinsig_benign}

# VEP ClinSig benign (to be excluded later)
benign_vep_clinsig <- grepl( "benign", variants.df$CLIN_SIG)
summary(benign_vep_clinsig)

# mean count of variants per case 
benign_vep_clinsig.mx <- genotypes.mx[benign_vep_clinsig,]
benign_vep_clinsig.counts <- apply(benign_vep_clinsig.mx, 2, sum, na.rm=TRUE)
hist(benign_vep_clinsig.counts, main="Counts of benign VEP ClinSig variants per individual")
benign_vep_clinsig.mean <- mean(benign_vep_clinsig.counts, na.rm=TRUE)
benign_vep_clinsig.mean
rm(benign_vep_clinsig.mx, benign_vep_clinsig.counts)

# inconsistent VEP ClinSig annotations
sum(vep_clinsig & benign_vep_clinsig) # will be excluded later

# Clean-up
rm(vep_clinsig_counts.df, benign_vep_clinsig.mean)

```

### Exclude benign or conflicting ClinVar ClnSig variants 

```{r inconsistent_clnsig}

# Exclude known benign or conflicting variants from any of the previously selected categiries
sum(lof)
sum(ddm)
sum(vep_clinsig)

lof <- lof & !benign_vep_clinsig
ddm <- ddm & !benign_vep_clinsig
vep_clinsig <- vep_clinsig & !benign_vep_clinsig

sum(lof)
sum(ddm)
sum(vep_clinsig)

rm(benign_vep_clinsig)

```

# Select protein-affecting variants

1,309 -> 162  

```{r select_variants}

# Make list of selected variants
selected_variants <- lof | vep_clinsig | ddm
summary(selected_variants)

```

# Explore dataset with protein-affected variants

## Barplot 

```{r selected_variants_barplot}

# --- Recalculate mean counts per case --- #
# After removal of known benign and contradictory variants  

# mean count of LoF-s per case
lof.mx <- genotypes.mx[lof,]
lof.counts <- apply(lof.mx, 2, sum, na.rm=TRUE)
table(lof.counts) # 10 cases with two low frequency LoF variants
lof.mean <- mean(lof.counts, na.rm=TRUE) 
lof.mean

# mean count of ddm per case
ddm.mx <- genotypes.mx[ddm,]
ddm.counts <- apply(ddm.mx, 2, sum, na.rm=TRUE)
table(ddm.counts) # 5 cases with multiple low frequency ddm variants
ddm.mean <- mean(ddm.counts, na.rm=TRUE)
ddm.mean

# mean count of clnsig per case 
clnsig.mx <- genotypes.mx[vep_clinsig,]
clnsig.counts <- apply(clnsig.mx, 2, sum, na.rm=TRUE)
table(clnsig.counts) # 1 case with 2 known pathogenic low frequency ClinSig variants
clnsig.mean <- mean(clnsig.counts, na.rm=TRUE)
clnsig.mean

# mean count of all selected variants per case 
all_selected.mx <- genotypes.mx[selected_variants,]
all_selected.counts <- apply(all_selected.mx, 2, sum, na.rm=TRUE)
table(all_selected.counts) # many cases with multiple candidate variants
all_selected.mean <- mean(all_selected.counts, na.rm=TRUE)
all_selected.mean

# Summary table with average numbers of protein-affecting per individual
summary_counts.df <- as.data.frame(cbind(
  type=c("all_selected", "ddm", "LoF", "clnsig"),
  count=c(all_selected.mean, ddm.mean, lof.mean, clnsig.mean)))

# Convert numbers to numeric vector
summary_counts.df$count <- as.numeric(as.vector(summary_counts.df$count))

# Order of levels defines order of bars on the plot
summary_counts.df$type <- factor(summary_counts.df$type, 
  levels=c("all_selected", "ddm", "LoF", "clnsig"))

# Plot bar diagramm of avg. count of selected types per individual
ggplot(summary_counts.df, aes(type, count)) +
  geom_bar(stat="identity", fill=c("red","blue","blue","blue"), alpha=0.5) + 
  geom_text(aes(label=round(count,2)), vjust=-1, size=5) +
  ylim(0,0.6) +
  labs(x = "", y = "") + 
  theme(axis.text.x = element_text(hjust=1, size=18, angle=45)) +
  ggtitle("Average counts of protein-affecting variants per individual\n(ampliseq-nfe, MAF<0.05)")

# Clean-up
rm(lof.mean, lof.mx, lof.counts, 
   ddm.mean, ddm.mx, ddm.counts, 
   clnsig.mean, clnsig.mx, clnsig.counts,
   all_selected.mean, all_selected.mx, all_selected.counts,
   summary_counts.df)

```

## Venn diagram

```{r venn_diagram}

# Count combinations of selectred categories
lof_ddm <- lof & ddm
clnsig_ddm <- vep_clinsig & ddm
lof_clnsig <- lof & vep_clinsig
lof_clnsig_ddm <- lof & vep_clinsig & ddm

summary(lof_ddm)
summary(clnsig_ddm)
summary(lof_clnsig)

# Plot venn diagram of selected variants by types
grid.newpage()
draw.triple.venn(
  area1=sum(lof), 
  area2=sum(vep_clinsig), 
  area3=sum(ddm), 
  n12=sum(lof_clnsig), 
  n23=sum(clnsig_ddm), 
  n13=sum(lof_ddm),
  n123=sum(lof_clnsig_ddm),
  category=c(
    paste("lof\n", sum(lof)),
    paste("clnsig\n", sum(vep_clinsig)),
    paste("ddm\n", sum(ddm))),
  fill = c("red", "green", "blue"), 
  alpha = c(0.3, 0.3, 0.3),
  cex=2)

# Clean-up
rm(lof, ddm, vep_clinsig, 
   clnsig_ddm, lof_clnsig, lof_ddm, lof_clnsig_ddm)

```

# Keep selected variants only  

```{r keep_only_selected}

# Keep selected variants only
variants.df <- variants.df[selected_variants,]
genotypes.mx <- genotypes.mx[selected_variants,]

# Check result
dim(variants.df)
dim(genotypes.mx)

# Clean-up
rm(selected_variants)

```

# Explore selected dataset

## Genotypes and missingness

```{r genotypes_missingness}

summary(as.factor(genotypes.mx))
sum(is.na(genotypes.mx))/(nrow(genotypes.mx)*ncol(genotypes.mx))

```

## multiallelic_varaints

Might violate some implicit intuitive statistical assumptions  

```{r count_selected_multiallelic_varaints}

sum(variants.df$MultiAllelic)

```

# explore_AFs_in_selected_variants

Most of selected variants are rare: 164 out of 169  
also the weihgting will further downgrade the variants with higher AF  

```{r explore_AFs_in_selected_variants}

mean(variants.df$total_AF)
median(variants.df$total_AF)

hist(variants.df$total_AF, main="AFs of the selected protein-affecting variants", labels = TRUE, ylim=c(0,180))

```

# explore_variants_counts_per_gene

Only 62 genes retained out of 107  
46 targeted genes had no low frequency protein-affecting variants  
1 gene was mixed-up (DCLRE1C was included instead of CDLRE1C)  

Still there are some new low-frtequency protein-affecting variants in BRCA1/2 and PALB2  
(after excluding the known pathogenic and benign ones)  

It is worth noting that some of the protein-affecting variants, previously detected in NFE (including variants in FOXM1 and SLC9A3R1) are not detected now.  Most likely this happened because of different variant filtering criteria that had to be applied in Ampliseq, comparatively to the previous WES analysis. Similar observation was previously made about the PALB2 carriers, which were detected in Ampliseq-only, but not in Ampliseq-NFE.  

```{r explore_variants_counts_per_gene}

# Get number of genes
length(unique(variants.df$SYMBOL))

# Genes not carrying any rare or low frequency protein-affecting variants
genes_62 <- unique(variants.df$SYMBOL)
missed_genes <- setdiff(genes_107, genes_62)
sort(missed_genes)
length(unique(missed_genes))

# Genes carrying at least one predicted low frequency protein-affecting variant
sort(genes_62)

# Counts of variants per gene
x <- sort(table(as.vector(variants.df$SYMBOL)), decreasing = TRUE)
x

# Table for histogram of numbers of protein-affecting variants per gene 
# (not sure whether group_by preserves the order of genes etc, 
# but this is not important here)
y <- variants.df %>% group_by(SYMBOL) %>% summarise(n())
"vars_count" -> colnames(y)[2]

# Plot the histogram
hist(y$vars_count, main="Aggregated AC-s of protein-affecting variants per gene\n(ampliseq-nfe, MAF<0.05)", xlab="AC per gene", ylab="Number of genes", ylim=c(0,40), breaks=c(1:12), labels = TRUE)

# Mean number of variants per gene is slightly higher than in the WES dataset (~4 vs ~2)
mean(y$vars_count)

# Clean-up
rm(x,y,genes_107,genes_62,missed_genes)

```

# Final check

```{r end_check}

dim(genotypes.mx)
dim(phenotypes.df)
dim(variants.df)

sum(rownames(variants.df) != rownames(genotypes.mx))
sum(colnames(genotypes.mx) != rownames(phenotypes.df))

```

# save_selected_variants_to_a_text_file

```{r write_text_file}

results_file <- paste(base_folder, "s02_selected_variants_162.txt", sep="/")
write.table(variants.df, file=results_file, quote=FALSE, sep="\t")
rm(results_file)

```

# save_results

```{r save_results}

save.image(paste(base_folder, "s02_filter_by_variant_effect.RData", sep="/"))

```

# final_section

```{r final_section}

ls()
sessionInfo()
Sys.time()

```
