---
title: "crude_genes_stats"
output: html_document
editor_options: 
  chunk_output_type: console
---

started: Alexey Larionov, 27Feb2017  
last updated: Alexey Larionov, 05Sep2019

# Summary  

164 variants -> 62 genes

Make a data frame with gene-level aggregated crude statistics:  

- Fisher tests for NFE vs BC  
- Fisher test for UBC vs CBC  
- prop.trend.test for NFE:UBC:CBC along with 1:2:3 scores  

### Explore crude aggregated stats per gene  
Mean agregated MAC: 5.8  
Median agregated MAC: 3  

qq-plots involving NFE (NFE-vs-BC and NFE-UBC-CBC trend) show p-values reasonably close to expected under H0.  
The deflation for UBC-vs-CBC may reflect low numbers of cases and low counts of variants with AF < 0.05  

###Input data:  
- 164 variants x 667 samples (470BC = 233UBC + 237CBC and 197NFE)  

###Output data:  
- 164 variants x 667 samples  
- 62 genes x 667 samples  

# start_section

<style>
pre {
  overflow-x: auto;
}
pre code {
  word-wrap: normal;
  white-space: pre;
}
</style>

```{r set_width, echo=F}

#eval=F

options(width = 999)
# https://stackoverflow.com/questions/36845178/width-of-r-code-chunk-output-in-rmarkdown-files-knitr-ed-to-html

```

```{r start_section}

# Time stamp
Sys.time()

# Clenan-up
rm(list=ls())
graphics.off()

# Base folder
library(knitr)
base_folder="/Users/alexey/Documents/wecare/ampliseq/v04_ampliseq_nfe/s17_crude_stats"
opts_knit$set(root.dir = base_folder)
#setwd(base_folder)

# Load libraries
library(dplyr)

# Accessory function
source(paste(base_folder, "f01_qqunif_plot.R", sep="/"))

```

# load_data

```{r load_data}

load(paste(base_folder, "s01_crude_variants_stats.RData", sep="/"))

```

# check_data

```{r check_data}

ls()

dim(genotypes.mx)
genotypes.mx[1:5,1:5]

dim(variants.df)
variants.df[1:5,1:5]

dim(phenotypes.df)
phenotypes.df[1:5,1:5]

# Check consistency of rownames and colnames
sum(colnames(genotypes.mx) != rownames(phenotypes.df))
sum(rownames(genotypes.mx) != rownames(variants.df))

```

# Get raw stats per gene

Record if multilaalelic variant is present in a gene.  
Calculate Fisher-exact tests and trend in proportion for raw counts.  
Note that trends in raw counts are evaluated **BEFORE** imputation.  
The result is still in character matrix that need to be converted to df with appropriate data types.  

```{r get_raw_stats_per_gene}

# Get list of genes 
# (check that there is no NAs; get rid of NAs if they are present!)
genes <- unique(as.character(variants.df$SYMBOL))
length(genes)
sum(is.na(genes))

# Prepare blank matrix for output
header <- c("gene", "num_var", "multiallelic", 
            "aggr_ac", "aggr_an", "aggr_af",  
            "aggr_ac_nfe", "aggr_an_nfe", "aggr_af_nfe",  
            "aggr_ac_bc", "aggr_an_bc", "aggr_af_bc",  
            "aggr_ac_cbc", "aggr_an_cbc", "aggr_af_cbc", 
            "aggr_ac_ubc", "aggr_an_ubc", "aggr_af_ubc",
            "nfe_bc_fisher_p", "ubc_cbc_fisher_p", "prop_trend_p", "crude_trend_direction")

genes.mx <- matrix(nrow=0, ncol=length(header))
header -> colnames(genes.mx)

# Make logical vectors for the subgroups
table(phenotypes.df$cc)
nfe_cases <- phenotypes.df$cc == -1
bc_cases <- phenotypes.df$cc != -1
ubc_cases <- phenotypes.df$cc == 0
cbc_cases <- phenotypes.df$cc == 1

#genes <- c("EPHB2","SZT2") # for testing

# For each gene
for(gene in genes){
  
  #gene="EPHB2" # 1 var # for testing
  #gene="SZT2" # 5 vars # for testing
  
  # Get list of variants
  vars <- variants.df$SYMBOL == gene

  # Count variants
  num_var <- sum(vars)

  # Check if any variant is multiallelic
  multiallelic <- (any(variants.df[vars, "Multiallelic"]))
  
  # Get genotypes matrices for subgroups
  gt.mx <- genotypes.mx[vars, , drop=FALSE]
  # drop=FALSE prevents converting data frame to vector for single-variant genes  
  # https://stat.ethz.ch/R-manual/R-devel/library/base/html/Extract.data.frame.html  
  
  gt_nfe.mx <- gt.mx[,nfe_cases, drop=FALSE]
  gt_bc.mx <- gt.mx[,bc_cases, drop=FALSE]
  gt_cbc.mx <- gt.mx[,cbc_cases, drop=FALSE]
  gt_ubc.mx <- gt.mx[,ubc_cases, drop=FALSE]
  
  # Calculate aggregated metrics in sub-groups
  aggr_ac <- sum(gt.mx, na.rm=TRUE)
  aggr_an <- 2*sum(!is.na(gt.mx))
  aggr_af <- aggr_ac / aggr_an

  aggr_ac_nfe <- sum(gt_nfe.mx, na.rm=TRUE)
  aggr_an_nfe <- 2*sum(!is.na(gt_nfe.mx))
  aggr_af_nfe <- aggr_ac_nfe / aggr_an_nfe
  
  aggr_ac_bc <- sum(gt_bc.mx, na.rm=TRUE)
  aggr_an_bc <- 2*sum(!is.na(gt_bc.mx))
  aggr_af_bc <- aggr_ac_bc / aggr_an_bc

  aggr_ac_cbc <- sum(gt_cbc.mx, na.rm=TRUE)
  aggr_an_cbc <- 2*sum(!is.na(gt_cbc.mx))
  aggr_af_cbc <- aggr_ac_cbc / aggr_an_cbc
  
  aggr_ac_ubc <- sum(gt_ubc.mx, na.rm=TRUE)
  aggr_an_ubc <- 2*sum(!is.na(gt_ubc.mx))
  aggr_af_ubc <- aggr_ac_ubc / aggr_an_ubc

  # Fisher exact test: nfe vs bc
  fisher_nfe_bc.mx <- matrix(c(aggr_ac_nfe, aggr_an_nfe - aggr_ac_nfe, aggr_ac_bc, aggr_an_bc - aggr_ac_bc), nrow=2)
  nfe_bc_fisher_p <- fisher.test(fisher_nfe_bc.mx)$p.value

  # Fisher exact test: cbc vs ubc
  fisher_cbc_ubc.mx <- matrix(c(aggr_ac_cbc, aggr_an_cbc - aggr_ac_cbc, aggr_ac_ubc, aggr_an_ubc - aggr_ac_ubc), nrow=2)
  cbc_ubc_fisher_p <- fisher.test(fisher_cbc_ubc.mx)$p.value

  # Proportions trend test: nfe-ubc-cbc
  ac_prop_trend <- c(aggr_ac_nfe, aggr_ac_ubc, aggr_ac_cbc)
  an_prop_trend <- c(aggr_an_nfe, aggr_an_ubc, aggr_an_cbc)
  prop_trend_p <- prop.trend.test(ac_prop_trend, an_prop_trend)$p.value
  
  # Trend call - to check WHY sum() ?
  crude_trend_direction <- "None"
  if(sum(aggr_af_nfe) < sum(aggr_af_cbc)) crude_trend_direction <- "Risk"
  if(sum(aggr_af_nfe) > sum(aggr_af_cbc)) crude_trend_direction <- "Protective"

  result <- c(gene, num_var, multiallelic, 
              aggr_ac, aggr_an, aggr_af, 
              aggr_ac_nfe, aggr_an_nfe, aggr_af_nfe, 
              aggr_ac_bc, aggr_an_bc, aggr_af_bc, 
              aggr_ac_cbc, aggr_an_cbc, aggr_af_cbc, 
              aggr_ac_ubc, aggr_an_ubc, aggr_af_ubc,
              nfe_bc_fisher_p, cbc_ubc_fisher_p, prop_trend_p, crude_trend_direction)
  
  # Record to result matrix 
  genes.mx <- rbind(genes.mx, result) 

  # Clean-up
  rm(vars, num_var, multiallelic, 
     gt.mx, gt_nfe.mx, gt_bc.mx, gt_cbc.mx, gt_ubc.mx, 
     aggr_ac, aggr_an, aggr_af, 
     aggr_ac_nfe, aggr_an_nfe, aggr_af_nfe, 
     aggr_ac_bc, aggr_an_bc, aggr_af_bc, 
     aggr_ac_cbc, aggr_an_cbc, aggr_af_cbc, 
     aggr_ac_ubc, aggr_an_ubc, aggr_af_ubc,
     fisher_nfe_bc.mx, fisher_cbc_ubc.mx, ac_prop_trend, an_prop_trend, 
     nfe_bc_fisher_p, cbc_ubc_fisher_p, prop_trend_p, crude_trend_direction, result)

}

# Check result
dim(genes.mx)

# Add rownames
genes -> rownames(genes.mx)

# Clean-up
rm(header, gene, genes, 
   nfe_cases, bc_cases, ubc_cases, cbc_cases)

```

# Convert chr stats mx to df with appropriate datatypes

```{r convert_stats_to_df}

genes.df <- as.data.frame(genes.mx, stringsAsFactors = FALSE)
str(genes.df)

genes.df$num_var <- as.integer(genes.df$num_var)

genes.df$multiallelic <- as.logical(genes.df$multiallelic)

genes.df$aggr_ac <-as.integer(genes.df$aggr_ac)
genes.df$aggr_an <-as.integer(genes.df$aggr_an)
genes.df$aggr_af <-as.numeric(genes.df$aggr_af)

genes.df$aggr_ac_nfe <-as.integer(genes.df$aggr_ac_nfe)
genes.df$aggr_an_nfe <-as.integer(genes.df$aggr_an_nfe)
genes.df$aggr_af_nfe <-as.numeric(genes.df$aggr_af_nfe)

genes.df$aggr_ac_bc <-as.integer(genes.df$aggr_ac_bc)
genes.df$aggr_an_bc <-as.integer(genes.df$aggr_an_bc)
genes.df$aggr_af_bc <-as.numeric(genes.df$aggr_af_bc)

genes.df$aggr_ac_cbc <-as.integer(genes.df$aggr_ac_cbc)
genes.df$aggr_an_cbc <-as.integer(genes.df$aggr_an_cbc)
genes.df$aggr_af_cbc <-as.numeric(genes.df$aggr_af_cbc)

genes.df$aggr_ac_ubc <-as.integer(genes.df$aggr_ac_ubc)
genes.df$aggr_an_ubc <-as.integer(genes.df$aggr_an_ubc)
genes.df$aggr_af_ubc <-as.numeric(genes.df$aggr_af_ubc)

genes.df$nfe_bc_fisher_p <-as.numeric(genes.df$nfe_bc_fisher_p)
genes.df$ubc_cbc_fisher_p <-as.numeric(genes.df$ubc_cbc_fisher_p)
genes.df$prop_trend_p <-as.numeric(genes.df$prop_trend_p)

# Check result
str(genes.df)

# Clean-up
rm(genes.mx)

```

# Add FDR-s

```{r add_fdrs}

nfe_bc_fisher_fdr <- p.adjust(genes.df$nfe_bc_fisher_p, method="fdr")
min(nfe_bc_fisher_fdr)

ubc_cbc_fisher_fdr <- p.adjust(genes.df$ubc_cbc_fisher_p, method="fdr")
min(ubc_cbc_fisher_fdr)

prop_trend_fdr <- p.adjust(genes.df$prop_trend_p, method="fdr")
min(prop_trend_fdr)

genes.df <- data.frame(genes.df,
                                 nfe_bc_fisher_fdr,
                                 ubc_cbc_fisher_fdr,
                                 prop_trend_fdr)

# Clean-up
rm(nfe_bc_fisher_fdr,ubc_cbc_fisher_fdr,prop_trend_fdr)

```

# Explore the raw stas

```{r explore_raw_stats}

# Show genes with multiallelic variants
summary(genes.df$multiallelic) # 2 genes with multiallelic variants
genes.df %>% 
  filter(multiallelic) %>% 
  select(gene)

# Look at the genes with significant crude trend
table(genes.df[,"crude_trend_direction"])
genes.df %>% 
  filter(prop_trend_p <= 0.05) %>% 
  select(gene, crude_trend_direction)

```

# explore_crude_aggr_afs_and_p_values

### Selected genes of interest

```{r selected_genes}

genes.df[c("ATM","CHEK2","NF1","NRG1","ERCC6"), 
                   c("aggr_af_nfe", "aggr_af_ubc", "aggr_af_cbc", 
                     "nfe_bc_fisher_p", "ubc_cbc_fisher_p", "prop_trend_p")]

```

Different p-values do not correlate to each other  

```{r p-values_against_each_other}

# Plot different p-values against each other
plot(genes.df$nfe_bc_fisher_p, genes.df$ubc_cbc_fisher_p)
plot(genes.df$nfe_bc_fisher_p, genes.df$prop_trend_p)
plot(genes.df$ubc_cbc_fisher_p, genes.df$prop_trend_p)

```

# Crude "risk" candidates

*SETX* still is much more present in WECARE-Ampliseq than in NFE  
Similarly, *BRIP1* is absent in NFE, but present in WECARE-Ampliseq  

```{r risk_candidates}

# Top "BC risk" candidates

genes.df %>% 
  filter(crude_trend_direction=="Risk", nfe_bc_fisher_p < 0.05) %>% 
  select(gene, aggr_af_nfe, aggr_af_ubc, aggr_af_cbc, nfe_bc_fisher_p, ubc_cbc_fisher_p, 
         prop_trend_p, nfe_bc_fisher_fdr, ubc_cbc_fisher_fdr, prop_trend_fdr) %>% 
  arrange(nfe_bc_fisher_p)

a <- genes.df %>% filter(crude_trend_direction=="Risk", nfe_bc_fisher_p < 0.05) %>% select(gene)

# Top "CBC risk" candidates
genes.df %>% 
  filter(crude_trend_direction=="Risk", ubc_cbc_fisher_p < 0.05) %>% 
  select(gene, aggr_af_nfe, aggr_af_ubc, aggr_af_cbc, nfe_bc_fisher_p, ubc_cbc_fisher_p, 
         prop_trend_p, nfe_bc_fisher_fdr, ubc_cbc_fisher_fdr, prop_trend_fdr) %>% 
  arrange(ubc_cbc_fisher_p)

b <- genes.df %>% filter(crude_trend_direction=="Risk", ubc_cbc_fisher_p < 0.05) %>% select(gene)

# Top "trend risk" candidates
genes.df %>% 
  filter(crude_trend_direction=="Risk", prop_trend_p < 0.05) %>% 
  select(gene, aggr_af_nfe, aggr_af_ubc, aggr_af_cbc, nfe_bc_fisher_p, ubc_cbc_fisher_p, 
         prop_trend_p, nfe_bc_fisher_fdr, ubc_cbc_fisher_fdr, prop_trend_fdr) %>% 
  arrange(prop_trend_p)

c <- genes.df %>% filter(crude_trend_direction=="Risk", prop_trend_p < 0.05) %>% select(gene)

# Intersects

intersect(a,b)
intersect(b,c)
intersect(a,c)

# Clean-up
rm(a,b,c)

```

# Crude "protective" candidates

Anyway: it does not make sence to look at "protective" in low-frequency,  
because  should not be rare, if beneficial ...  

```{r protective_candidates}

# Top "BC protective" candidates
genes.df %>% 
  filter(crude_trend_direction=="Protective", nfe_bc_fisher_p < 0.05) %>% 
  select(gene, aggr_af_nfe, aggr_af_ubc, aggr_af_cbc, nfe_bc_fisher_p, ubc_cbc_fisher_p, 
         prop_trend_p, nfe_bc_fisher_fdr, ubc_cbc_fisher_fdr, prop_trend_fdr) %>% 
  arrange(nfe_bc_fisher_p)

a <- genes.df %>% filter(crude_trend_direction=="Protective", nfe_bc_fisher_p < 0.05) %>% select(gene)

# Top "CBC protective" candidates
genes.df %>% 
  filter(crude_trend_direction=="Protective", ubc_cbc_fisher_p < 0.05) %>% 
  select(gene, aggr_af_nfe, aggr_af_ubc, aggr_af_cbc, nfe_bc_fisher_p, ubc_cbc_fisher_p, 
         prop_trend_p, nfe_bc_fisher_fdr, ubc_cbc_fisher_fdr, prop_trend_fdr) %>% 
  arrange(ubc_cbc_fisher_p)

b <- genes.df %>% filter(crude_trend_direction=="Protective", ubc_cbc_fisher_p < 0.05) %>% select(gene)

# Top "trend protective" candidates
genes.df %>% 
  filter(crude_trend_direction=="Protective", prop_trend_p < 0.05) %>% 
  select(gene, aggr_af_nfe, aggr_af_ubc, aggr_af_cbc, nfe_bc_fisher_p, ubc_cbc_fisher_p, 
         prop_trend_p, nfe_bc_fisher_fdr, ubc_cbc_fisher_fdr, prop_trend_fdr) %>% 
  arrange(prop_trend_p)

c <- genes.df %>% filter(crude_trend_direction=="Protective", prop_trend_p < 0.05) %>% select(gene)

# DCHS2 and GBGT1 are "Protective" because of the NFE vs WECARE change

# Intersects
intersect(a,b)
intersect(b,c)
intersect(a,c)

# Clean-ups
rm(a,b,c)

```

# QQ-plots

Assuming uniform p-value distributions under the null-hypothesis  

```{r qq-plots}

# nfe_bc_fisher_p
min(genes.df$nfe_bc_fisher_p)

qqunif.plot(genes.df[,"nfe_bc_fisher_p"], 
            main="NFE vs BC\ncrude counts aggregated per gene\nFisher exact test")

# ubc_cbc_fisher_p
min(genes.df$ubc_cbc_fisher_p)

qqunif.plot(genes.df$ubc_cbc_fisher_p, 
            main="UBC vs CBC\ncrude counts aggregated per gene\nFisher exact test")

# prop trend p
min(genes.df$prop_trend_p)

qqunif.plot(genes.df$prop_trend_p, 
            main="NFE-UBC-CBC\ncrude counts aggregated per gene\nTrend in proportions test")

# Clean-up
rm(qqunif.plot)

```

# explore_MACs

Crude allelic counts for ATM and CHEK2 are higher in BC than in NFE  
However, no further increase in CBC vs UBC (rather contrary for CHEK2)  

```{r explore_MACs}

# Selected genes: candidates from previous analyses and
# GFRAL was an outlier, because of NFE >> UBC/CBC : likely methodical difference between KGEN and WECARE
genes.df[c("ATM","CHEK2"), 
                   c("aggr_ac", "aggr_ac_nfe", "aggr_ac_bc", "aggr_ac_ubc", "aggr_ac_cbc")]

# Histograms of aggregated MACs
hist(genes.df$aggr_ac, labels = TRUE, ylim=c(0,65), 
     xlab="aggregated MACs per gene (bins)", ylab="num of genes")

hist(genes.df$aggr_ac[genes.df$aggr_ac<26], breaks=0:25, 
     labels = TRUE, ylim=c(0,20), main="zoom to MAC < 25", 
     xlab="aggregated MACs per gene", ylab="num of genes")

mean(genes.df$aggr_ac)
median(genes.df$aggr_ac)

# Genes with high MACs (non-weighted!)
x <- genes.df %>% 
  select(gene, num_var, multiallelic, aggr_ac, nfe_bc_fisher_p, ubc_cbc_fisher_p,crude_trend_direction) %>% 
  arrange(desc(aggr_ac))
x[1:10,]

# Aggregated counts < 10 - too rare for stat analysis ??
sum(genes.df$aggr_ac < 10)

# Genes that may be analysed
sum(genes.df$aggr_ac >= 10)

# clean-up
rm(x)

```

# data_summary

```{r data_summary}

ls()

dim(genotypes.mx)
dim(variants.df)
dim(phenotypes.df)

dim(genes.df)

# Check consistency of rownames and colnames
sum(colnames(genotypes.mx) != rownames(phenotypes.df))
sum(rownames(genotypes.mx) != rownames(variants.df))

```

# save_data

```{r save_data}

save.image(paste(base_folder, "s02_crude_genes_stats.RData", sep="/"))

```

# final_section

```{r final_section}

ls()
sessionInfo()
Sys.time()

```
