---
title: "s01_joined_PCA_plots_971_not_in_LD"
output: html_document
editor_options: 
  chunk_output_type: console
---

started: AL26Apr2019  
last updated: AL14Jul2019  

# Summary  

Input sequencing data: 1,850 vars x 715 samples ( 519BC = 258UBC + 260CBC and 197NFE)  
Input eigenvectors: 3,219 samples = 715 wecare-nfe + 2,504 kgen 

# Start_section

<style>
pre {
  overflow-x: auto;
}
pre code {
  word-wrap: normal;
  white-space: pre;
}
</style>

```{r set_width, echo=F}

#eval=F

options(width = 999)
# https://stackoverflow.com/questions/36845178/width-of-r-code-chunk-output-in-rmarkdown-files-knitr-ed-to-html

```

```{r start_section}

Sys.time()
rm(list=ls())
graphics.off()

library(knitr)
library(dplyr)
library(ggplot2)
library(plotly)

base_folder="/Users/alexey/Documents/wecare/ampliseq/v04_ampliseq_nfe/s14_joined_ampliseq_1kg_PCs/s03_all_variants_not_in_LD_971"

opts_knit$set(root.dir = base_folder)

options(stringsAsFactors = F)
options(warnPartialMatchArgs = T, 
        warnPartialMatchAttr = T, 
        warnPartialMatchDollar = T)

#options(error = browser()) # Type Q or c to exit, drop browser level
# https://support.rstudio.com/hc/en-us/articles/200713843?version=1.1.456&mode=desktop
# https://stackoverflow.com/questions/13052522/how-to-leave-the-r-browser-mode-in-the-console-window/13052588 

```

# Read_data  

```{r read_data}

# Sequencing data (for wecare phenotypes: case/control status)
source_folder="/Users/alexey/Documents/wecare/ampliseq/v04_ampliseq_nfe/s12_check_BRCA1_BRCA2_PALB2"
load(paste(source_folder, "s03_exclude_BRCA1_BCRA2_PALB2_carriers.RData", sep="/"))
base_folder="/Users/alexey/Documents/wecare/ampliseq/v04_ampliseq_nfe/s14_joined_ampliseq_1kg_PCs/s03_all_variants_not_in_LD_971"

# Eigenvectors
eigenvectors_file <- paste(base_folder, "ampliseq_1kg_971_3219_100PCs.eigenvec", sep="/")
eigenvectors.df <- read.table(eigenvectors_file, header=T, sep="\t",quote="")

# Eigenvalues
eigenvalues_file <- paste(base_folder, "ampliseq_1kg_971_3219_100PCs.eigenval", sep="/")
eigenvalues.df <- read.table(eigenvalues_file, header=F, sep="\t",quote="")

# 1kg phenotypes (ethnicity and gender)
kg_phenotypes_file <- paste(base_folder, "integrated_call_samples_v3.20130502.ALL.panel", sep="/")
kg_phenotypes.df <- read.table(kg_phenotypes_file, header=T)

# Clean-up
rm(source_folder, eigenvectors_file, eigenvalues_file, kg_phenotypes_file, genotypes.mx, variants.df)

```

# Check data

```{r start_check}

# List of objects
ls()

# Expected number of samples in eigenvectors
2504+715

# Dimentions of objects
dim(eigenvectors.df)
dim(eigenvalues.df)
dim(kg_phenotypes.df)
dim(phenotypes.df)

# Update eigenvectors
rownames(eigenvectors.df) <- eigenvectors.df$FID
eigenvectors.df <- eigenvectors.df[,-1]

```

# Plot eigenvalues

```{r eigenvalues}

plot(eigenvalues.df$V1, type="b", ylab="Variance",
     main="Top 100 eigenvectors")

plot(eigenvalues.df$V1[1:10], type="b", ylab="Variance",
     main="Top 10 eigenvectors")

rm(eigenvalues.df)

```

# Prepare data for Ampliseq-Kgen plot

Overlapping NFE samples should be removed from Amplise data before plotting  

```{r ampliseq_kgen_data}

# Make a table with IDs of overlapping NFE
eigenvectors.df[c(3022,3023),c("IID","PC1")]
nfe_pca <- eigenvectors.df$IID[3023:3219] # re-processed NFE added to Ampliseq
nfe_ampliseq <- sub("2:","",nfe_pca)
nfe.df <- data.frame(nfe_ampliseq, nfe_pca)

# Remove overlapping NFE from ampliseq-kgen eigenvectors
selected_samples <- ! eigenvectors.df$IID %in% nfe.df$nfe_pca
sum(selected_samples)
518+2504
eigenvectors_ampliseq_kgen.df <- eigenvectors.df[selected_samples,1:6]
"sample" -> colnames(eigenvectors_ampliseq_kgen.df)[1]

# Prepare ampliseq phenotypes 
# (for clarity of PCA plot wecare samples and controls were not separated)
phenotypes.df[c(518:519),c(1,2)]
phenotypes_ampliseq.df <- phenotypes.df[1:518,c("long_ids","cc")]
table(phenotypes_ampliseq.df$cc)

"WECARE" -> phenotypes_ampliseq.df$cc[phenotypes_ampliseq.df$cc==1] # This could be named as WECARE-samples
"WECARE" -> phenotypes_ampliseq.df$cc[phenotypes_ampliseq.df$cc==0] # This could be named as WECARE-controls
table(phenotypes_ampliseq.df$cc)
c("sample","group") -> colnames(phenotypes_ampliseq.df)

# Prepare kgen phenotypes
phenotypes_kgen.df <- kg_phenotypes.df[,c("sample","super_pop")]
c("sample","group") -> colnames(phenotypes_kgen.df)

# Merge kgen and Ampliseq phenotypes (latest will be on top in the plot)
phenotypes_ampliseq_kgen.df <- rbind(phenotypes_kgen.df,phenotypes_ampliseq.df)
table(phenotypes_ampliseq_kgen.df$group)

# Add eigenvectors to phenotypes
eigenphen_ampliseq_kgen.df <- full_join(
  phenotypes_ampliseq_kgen.df, eigenvectors_ampliseq_kgen.df, by="sample")
dim(eigenphen_ampliseq_kgen.df)
head(eigenphen_ampliseq_kgen.df)
tail(eigenphen_ampliseq_kgen.df)

# Clean-up
rm(nfe_pca, nfe_ampliseq, selected_samples, eigenvectors_ampliseq_kgen.df, phenotypes_ampliseq.df,
   phenotypes_ampliseq_kgen.df, nfe.df, phenotypes_kgen.df)

```

# Make Ampliseq-Kgen PCA plot

```{r ampliseq_kgen_plot}

# Prepare vector fr colour scale
myColours <- c("EUR"="BLUE", "AFR"="YELLOW", "AMR"="GREEN",
               "SAS"="GREY", "EAS"="PINK", 
               "WECARE"="RED")

myColourScale <- scale_colour_manual(values=myColours)

# Static plot
ggplot(eigenphen_ampliseq_kgen.df, aes(PC1,PC2)) + 
  geom_point(aes(col=group)) +
  labs(title="971 variants not in LD", x="PC1", y="PC2") +
  myColourScale

# Interactive plot
plotly_group <- factor(eigenphen_ampliseq_kgen.df$group,
  levels=c("AFR","AMR","EAS","SAS","EUR","WECARE"))

g <- ggplot(eigenphen_ampliseq_kgen.df, aes(PC1,PC2)) + 
  geom_point(aes(col=plotly_group, text=sample)) +
  labs(title="971 variants not in LD", x="PC1", y="PC2") +
  theme(legend.title=element_blank()) + # To suppress the legend title
  myColourScale                         # otherwise it would be "plotly_group"

ggplotly(g, tooltip="text") # By default the tooltip would also show coordinates 

# Clean-up
rm(myColours, myColourScale, g)

```

# Prepare data for NFE-Kgen plot

The NFE subset in Ampliseq-nfe data had been re-aligned and re-called, so it would be interesting to see  
how it is positioned relatively to the same samples aligned/called as in the "original" kgen VCF  

```{r nfe_kgen_data}

# Select samples
eigenvectors.df[c(2504,2505),c("IID","PC1")]
eigenvectors.df[c(3022,3023),c("IID","PC1")]

eigenvectors_nfe_kgen.df <- eigenvectors.df[c(1:2504,3023:3219),1:6]
dim(eigenvectors_nfe_kgen.df)
2504+197
"sample" -> colnames(eigenvectors_nfe_kgen.df)[1]

# Prepare NFE phenotypes 
eigenvectors_nfe_kgen.df[2504:2505,"sample"]
sample <- eigenvectors_nfe_kgen.df[2505:2701,"sample"]
group <- rep("Re-processed_NFE",197)
phenotypes_nfe.df <- data.frame(sample,group)

# Prepare kgen phenotypes
phenotypes_kgen.df <- kg_phenotypes.df[,c("sample","super_pop")]
c("sample","group") -> colnames(phenotypes_kgen.df)

# Merge kgen and re-processed NFE phenotypes (latest will be on top in the plot)
phenotypes_nfe_kgen.df <- rbind(phenotypes_kgen.df,phenotypes_nfe.df)
table(phenotypes_nfe_kgen.df$group)

# Add eigenvectors to phenotypes
eigenphen_nfe_kgen.df <- full_join(
  phenotypes_nfe_kgen.df, eigenvectors_nfe_kgen.df, by="sample")
dim(eigenphen_nfe_kgen.df)
head(eigenphen_nfe_kgen.df)
tail(eigenphen_nfe_kgen.df)

# Clean-up
rm(sample, group, eigenvectors_nfe_kgen.df, phenotypes_nfe.df,
   phenotypes_nfe_kgen.df, phenotypes_kgen.df)

```

# Make NFE-Kgen PCA plot

```{r nfe_kgen_plot}

# Prepare vector fr colour scale
myColours <- c("EUR"="BLUE", "AFR"="YELLOW", "AMR"="GREEN",
               "SAS"="GREY", "EAS"="PINK", 
               "Re-processed_NFE"="CYAN")

myColourScale <- scale_colour_manual(values=myColours)

# Static plot
ggplot(eigenphen_nfe_kgen.df, aes(PC1,PC2)) + 
  geom_point(aes(col=group)) +
  labs(title="Re-processed NFE", x="PC1", y="PC2") +
  myColourScale

# Interactive plot
plotly_group <- factor(eigenphen_nfe_kgen.df$group,
  levels=c("AFR","AMR","EAS","SAS","EUR","Re-processed_NFE"))

g <- ggplot(eigenphen_nfe_kgen.df, aes(PC1,PC2)) + 
  geom_point(aes(col=plotly_group, text = sample)) +
  labs(title="Re-processed NFE", x="PC1", y="PC2") +
  theme(legend.title=element_blank()) + # To suppress the legend title
  myColourScale                         # otherwise it would be "plotly_group"

ggplotly(g, tooltip="text") # By default the tooltip would also show coordinates 

# Clean-up
rm(myColours, myColourScale, g)

```

# Final_section

```{r final_section}

ls()
sessionInfo()
Sys.time()

```
